File: internal/models/deploy/deploy.go
// Package deploy 定义部署流程相关的数据模型
// 本文件包含部署记录相关的模型定义
package deploy

import (
	"time"

	"gorm.io/gorm"
)

// ==================== 部署记录模型 ====================

// DeployRecord 部署记录（含审批流程）
// 记录每次部署的详细信息
type DeployRecord struct {
	ID            uint      `gorm:"primarykey" json:"id"`
	CreatedAt     time.Time `json:"created_at"`
	UpdatedAt     time.Time `json:"updated_at"`
	ApplicationID uint      `gorm:"not null;index" json:"application_id"`          // 应用ID
	AppName       string    `gorm:"size:100;index" json:"app_name"`                 // 应用名称
	EnvName       string    `gorm:"size:50;index" json:"env_name"`                  // 环境名称
	Version       string    `gorm:"size:100" json:"version"`                        // 版本号
	Branch        string    `gorm:"size:100" json:"branch"`                         // Git 分支
	CommitID      string    `gorm:"size:100" json:"commit_id"`                      // Git Commit ID
	ImageTag      string    `gorm:"size:200" json:"image_tag"`                      // 镜像标签
	DeployType    string    `gorm:"size:50" json:"deploy_type"`                     // 部署类型: deploy/rollback/restart/scale
	DeployMethod  string    `gorm:"size:50;default:'jenkins'" json:"deploy_method"` // 部署方式: jenkins/k8s
	Status        string    `gorm:"size:20;index" json:"status"`                    // 状态: pending/approved/rejected/running/success/failed/cancelled
	Description   string    `gorm:"type:text" json:"description"`                   // 发布说明
	// Jenkins 相关
	JenkinsBuild int    `gorm:"default:0" json:"jenkins_build"` // Jenkins 构建号
	JenkinsURL   string `gorm:"size:500" json:"jenkins_url"`    // Jenkins 构建 URL
	// 审批相关
	NeedApproval    bool       `gorm:"default:false" json:"need_approval"` // 是否需要审批
	ApprovalChainID *uint      `gorm:"index" json:"approval_chain_id"`     // 审批链ID
	ApproverID      *uint      `json:"approver_id"`                        // 审批人ID
	ApproverName    string     `gorm:"size:100" json:"approver_name"`      // 审批人名称
	ApprovedAt      *time.Time `json:"approved_at"`                        // 审批时间
	RejectReason    string     `gorm:"type:text" json:"reject_reason"`     // 拒绝原因
	// 执行相关
	Duration     int        `gorm:"default:0" json:"duration"`  // 执行时长(秒)
	ErrorMsg     string     `gorm:"type:text" json:"error_msg"` // 错误信息
	Operator     string     `gorm:"size:100" json:"operator"`   // 操作人
	OperatorID   uint       `gorm:"index" json:"operator_id"`   // 操作人ID
	StartedAt    *time.Time `json:"started_at"`                 // 开始时间
	FinishedAt   *time.Time `json:"finished_at"`                // 完成时间
	RollbackFrom *uint      `json:"rollback_from"`              // 回滚来源记录ID
}

// TableName 指定表名
func (DeployRecord) TableName() string {
	return "deploy_records"
}

// DeployLock 发布锁
// 防止同一应用同一环境并发部署
type DeployLock struct {
	ID            uint       `gorm:"primarykey" json:"id"`
	CreatedAt     time.Time  `json:"created_at"`
	ApplicationID uint       `gorm:"not null;index" json:"application_id"`   // 应用ID
	EnvName       string     `gorm:"size:50;not null" json:"env_name"`       // 环境名称
	RecordID      uint       `gorm:"not null;index" json:"record_id"`        // 关联的部署记录ID
	LockedBy      uint       `gorm:"not null" json:"locked_by"`              // 锁定者ID
	LockedByName  string     `gorm:"size:100" json:"locked_by_name"`         // 锁定者名称
	ExpiresAt     time.Time  `gorm:"not null" json:"expires_at"`             // 过期时间
	Status        string     `gorm:"size:20;default:'active'" json:"status"` // 状态: active/released/expired
	ReleasedAt    *time.Time `json:"released_at"`                            // 释放时间
	ReleasedBy    *uint      `json:"released_by"`                            // 释放者ID
	ReleaseReason string     `gorm:"size:200" json:"release_reason"`         // 释放原因
}

// TableName 指定表名
func (DeployLock) TableName() string {
	return "deploy_locks"
}

// DeployWindow 发布窗口
// 定义允许发布的时间窗口
type DeployWindow struct {
	ID             uint      `gorm:"primarykey" json:"id"`
	CreatedAt      time.Time `json:"created_at"`
	UpdatedAt      time.Time `json:"updated_at"`
	AppID          uint      `gorm:"index;default:0" json:"app_id"`               // 应用ID，0表示全局
	Env            string    `gorm:"size:50;not null" json:"env"`                 // 环境
	Weekdays       string    `gorm:"size:50;default:'1,2,3,4,5'" json:"weekdays"` // 允许的星期: 1-7
	StartTime      string    `gorm:"size:10;default:'10:00'" json:"start_time"`   // 开始时间
	EndTime        string    `gorm:"size:10;default:'18:00'" json:"end_time"`     // 结束时间
	AllowEmergency bool      `gorm:"default:true" json:"allow_emergency"`         // 是否允许紧急发布
	Enabled        bool      `gorm:"default:true" json:"enabled"`                 // 是否启用
	CreatedBy      uint      `gorm:"default:0" json:"created_by"`
}

// TableName 指定表名
func (DeployWindow) TableName() string {
	return "deploy_windows"
}

// Task 任务模型
// 存储异步任务信息
type Task struct {
	gorm.Model
	Name        string    `gorm:"size:100;not null" json:"name"`                    // 任务名称
	Description string    `gorm:"type:text" json:"description"`                     // 描述
	Status      string    `gorm:"size:20;default:'pending';not null" json:"status"` // 状态
	CreatedBy   uint      `gorm:"not null" json:"created_by"`                       // 创建者ID
	StartTime   time.Time `json:"start_time"`                                       // 开始时间
	EndTime     time.Time `json:"end_time"`                                         // 结束时间
	JenkinsJob  string    `gorm:"size:100" json:"jenkins_job"`                      // Jenkins Job
	Parameters  string    `gorm:"type:text" json:"parameters"`                      // 参数 JSON
}

// TableName 指定表名
func (Task) TableName() string {
	return "tasks"
}

File: web/src/services/deploy.ts
import request from './api'
import type { ApiResponse } from '../types'

// 部署记录（含审批）
export interface DeployRecord {
  id: number
  created_at: string
  updated_at: string
  application_id: number
  app_name: string
  env_name: string
  version: string
  branch: string
  commit_id: string
  image_tag: string
  deploy_type: string
  deploy_method: string
  description: string
  status: string
  need_approval: boolean
  approver_id?: number
  approver_name?: string
  approved_at?: string
  reject_reason?: string
  jenkins_build: number
  jenkins_url: string
  duration: number
  error_msg?: string
  operator: string
  operator_id: number
  started_at?: string
  finished_at?: string
  rollback_from?: number
}

// 发布锁
export interface DeployLock {
  id: number
  created_at: string
  application_id: number
  env_name: string
  record_id: number
  locked_by: number
  locked_by_name: string
  expires_at: string
  status: string
  released_at?: string
  released_by?: number
  release_reason?: string
}

// 审批记录
export interface ApprovalRecord {
  id: number
  created_at: string
  record_id: number
  approver_id: number
  approver_name: string
  action: string
  comment?: string
}

// 部署统计
export interface DeployStats {
  total: number
  success: number
  failed: number
  success_rate: number
  avg_duration: number
}

// 创建部署参数
export interface CreateDeployDTO {
  application_id: number
  env_name: string
  version?: string
  branch?: string
  commit_id?: string
  image_tag?: string
  deploy_method?: string  // jenkins, k8s
  description?: string
}

// 回滚参数
export interface RollbackDTO {
  application_id: number
  env_name: string
}

export const deployApi = {
  // 部署记录
  createDeploy: (data: CreateDeployDTO): Promise<ApiResponse<DeployRecord>> => {
    return request.post('/deploy/records', data)
  },

  listRecords: (params?: {
    page?: number
    page_size?: number
    application_id?: number
    app_name?: string
    env_name?: string
    status?: string
  }): Promise<ApiResponse<{ list: DeployRecord[]; total: number }>> => {
    return request.get('/deploy/records', { params })
  },

  getRecord: (id: number): Promise<ApiResponse<{
    record: DeployRecord
    approvals: ApprovalRecord[]
  }>> => {
    return request.get(`/deploy/records/${id}`)
  },

  cancelDeploy: (id: number): Promise<ApiResponse> => {
    return request.post(`/deploy/records/${id}/cancel`)
  },

  approveDeploy: (id: number, comment?: string): Promise<ApiResponse> => {
    return request.post(`/deploy/records/${id}/approve`, { comment })
  },

  rejectDeploy: (id: number, reason: string): Promise<ApiResponse> => {
    return request.post(`/deploy/records/${id}/reject`, { reason })
  },

  executeDeploy: (id: number): Promise<ApiResponse> => {
    return request.post(`/deploy/records/${id}/execute`)
  },

  // 回滚
  createRollback: (data: RollbackDTO): Promise<ApiResponse<DeployRecord>> => {
    return request.post('/deploy/rollback', data)
  },

  getAvailableRollback: (appId: number, env: string): Promise<ApiResponse<DeployRecord>> => {
    return request.get(`/deploy/rollback/${appId}/${env}/available`)
  },

  // 锁管理
  getLockStatus: (appId: number, env: string): Promise<ApiResponse<{ locked: boolean; lock?: DeployLock }>> => {
    return request.get(`/deploy/locks/${appId}/${env}`)
  },

  releaseLock: (appId: number, env: string, reason?: string): Promise<ApiResponse> => {
    return request.post(`/deploy/locks/${appId}/${env}/release`, { reason })
  },

  // 统计
  getStats: (params?: {
    application_id?: number
    env_name?: string
    start_time?: string
    end_time?: string
  }): Promise<ApiResponse<DeployStats>> => {
    return request.get('/deploy/stats', { params })
  }
}


// ==================== 部署前置检查 ====================

export interface PreCheckItem {
  name: string
  status: 'passed' | 'warning' | 'failed' | 'skipped'
  message: string
  detail?: string
}

export interface DeployPreCheckRequest {
  application_id: number
  env_name: string
  image_tag?: string
}

export interface DeployPreCheckResponse {
  can_deploy: boolean
  checks: PreCheckItem[]
  warnings: string[]
  errors: string[]
}

// ==================== 灰度发布 ====================

export interface CanaryDeployRequest {
  application_id: number
  env_name: string
  image_tag: string
  canary_percent: number
  canary_replicas?: number
  description?: string
}

export interface CanaryStatus {
  deploy_record_id: number
  status: string
  canary_replicas: number
  stable_replicas: number
  canary_ready: number
  stable_ready: number
  canary_image: string
  stable_image: string
  started_at: string
  canary_healthy: boolean
  error_rate?: string
}

export interface CanaryRollbackRequest {
  deploy_record_id: number
  reason?: string
}

// ==================== 自动回滚 ====================

export interface AutoRollbackConfig {
  enabled: boolean
  health_check_period: number
  failure_threshold: number
  success_threshold: number
}

export interface DeployHealthStatus {
  deploy_record_id: number
  status: string
  ready_replicas: number
  desired_replicas: number
  unavailable_count: number
  restart_count: number
  last_check_time: string
  consecutive_fails: number
  should_rollback: boolean
  rollback_reason?: string
}

export const deployCheckApi = {
  // 部署前置检查
  preCheck: (data: DeployPreCheckRequest): Promise<ApiResponse<DeployPreCheckResponse>> => {
    return request.post('/deploy/pre-check', data)
  },

  // 灰度发布
  startCanary: (data: CanaryDeployRequest): Promise<ApiResponse<CanaryStatus>> => {
    return request.post('/deploy/canary/start', data)
  },

  getCanaryStatus: (recordId: number): Promise<ApiResponse<CanaryStatus>> => {
    return request.get(`/deploy/canary/${recordId}/status`)
  },

  promoteCanary: (recordId: number): Promise<ApiResponse<void>> => {
    return request.post(`/deploy/canary/${recordId}/promote`)
  },

  rollbackCanary: (recordId: number, reason?: string): Promise<ApiResponse<void>> => {
    return request.post(`/deploy/canary/${recordId}/rollback`, { reason })
  },

  // 自动回滚
  getDeployHealth: (recordId: number): Promise<ApiResponse<DeployHealthStatus>> => {
    return request.get(`/deploy/auto-rollback/${recordId}/health`)
  },

  updateRollbackConfig: (recordId: number, config: AutoRollbackConfig): Promise<ApiResponse<void>> => {
    return request.post(`/deploy/auto-rollback/${recordId}/config`, config)
  }
}
